src/goto-symex/build_goto_trace.cpp

```diff
@@ -248,97 +260,94 @@ void build_goto_trace(
+        goto_trace_step.lhs_object=
+          ssa_exprt(SSA_step.ssa_lhs.get_original_expr());
       else
+        goto_trace_step.lhs_object.make_nil();
+      goto_trace_step.type=SSA_step.type;
+      goto_trace_step.hidden=SSA_step.hidden;
+      goto_trace_step.format_string=SSA_step.format_string;
+      goto_trace_step.io_id=SSA_step.io_id;
+      goto_trace_step.formatted=SSA_step.formatted;
+      goto_trace_step.identifier=SSA_step.identifier;
+
+      // update internal field for specific variables in the counterexample
+      update_internal_field(SSA_step, goto_trace_step, ns);
+
+      goto_trace_step.assignment_type=
+        (SSA_step.is_assignment() &&
+         (SSA_step.assignment_type==
+          symex_targett::assignment_typet::VISIBLE_ACTUAL_PARAMETER ||
+          SSA_step.assignment_type==
+          symex_targett::assignment_typet::HIDDEN_ACTUAL_PARAMETER))?
+        goto_trace_stept::assignment_typet::ACTUAL_PARAMETER:
+        goto_trace_stept::assignment_typet::STATE;
+
+      if(SSA_step.original_full_lhs.is_not_nil())
+        goto_trace_step.full_lhs=
+          build_full_lhs_rec(
+            prop_conv, ns, SSA_step.original_full_lhs, SSA_step.ssa_full_lhs);
+
+      if(SSA_step.ssa_lhs.is_not_nil())
+        goto_trace_step.lhs_object_value=prop_conv.get(SSA_step.ssa_lhs);
+
+      if(SSA_step.ssa_full_lhs.is_not_nil())
       {
-        exprt tmp=prop_conv.get(j);
-        goto_trace_step.io_args.push_back(tmp);
+        goto_trace_step.full_lhs_value=prop_conv.get(SSA_step.ssa_full_lhs);
+        simplify(goto_trace_step.full_lhs_value, ns);
       }
-    }
 
-    if(SSA_step.is_assert() ||
-       SSA_step.is_assume() ||
-       SSA_step.is_goto())
-    {
-      goto_trace_step.cond_expr=SSA_step.cond_expr;
+      for(const auto &j : SSA_step.converted_io_args)
+      {
+        if(j.is_constant() ||
+           j.id()==ID_string_constant)
+          goto_trace_step.io_args.push_back(j);
+        else
+        {
+          exprt tmp=prop_conv.get(j);
+          goto_trace_step.io_args.push_back(tmp);
+        }
+      }
 
-      goto_trace_step.cond_value=
-        prop_conv.l_get(SSA_step.cond_literal).is_true();
-    }
-  }
+      if(SSA_step.is_assert() ||
+         SSA_step.is_assume() ||
+         SSA_step.is_goto())
+      {
+        goto_trace_step.cond_expr=SSA_step.cond_expr;
 
-  // Now assemble into a single goto_trace.
-  // This exploits sorted-ness of the map.
-  for(auto &t_it : time_map)
-    goto_trace.steps.splice(goto_trace.steps.end(), t_it.second);
+        goto_trace_step.cond_value=
+          prop_conv.l_get(SSA_step.cond_literal).is_true();
+      }
 
-  // cut off the trace at the desired end
-  for(goto_tracet::stepst::iterator
-      s_it1=goto_trace.steps.begin();
-      s_it1!=goto_trace.steps.end();
-      ++s_it1)
-    if(end_step_seen && end_ptr==&(*s_it1))
-    {
-      goto_trace.trim_after(s_it1);
-      break;
+      if(ssa_step_it == last_step_in_goto_trace)
```

Comment:
I think you are mixing up goto trace steps and SSA steps here. The trimming of the goto trace must happen wrt the recorded goto trace step of the violated assertion after having put everything in the right sequence.

Reply:
Yeah, that's probably what it wanted to do, but see this comment: https://github.com/diffblue/cbmc/pull/2610#issuecomment-407732786

Reply:
I think the fix for this is to change the meaning of 'end_step' to (none|last_step_to_keep). Then we can add a precondition to the build_goto_trace variant that the last_step_to_keep is an assertion (which is true for all existing users of that method). Cutting after the goto step corresponding to that last_step_to_keep should then give us the desired behaviour.

Reply:
Essentially, we need two variants: 1. Trim after first assertion, 2. Trim after _this_ assertion. To implement 1. we have to trim before concretising.